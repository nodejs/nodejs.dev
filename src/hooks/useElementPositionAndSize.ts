import { useState, useEffect } from 'react';

export interface ElementPositionAndSize {
  x: number;
  y: number;
  width: number;
  height: number;
}

export const useElementPositionAndSize = <T extends HTMLElement>(
  elementRef: React.RefObject<T>
) => {
  const [position, setPosition] = useState<ElementPositionAndSize>({
    x: 0,
    y: 0,
    width: 0,
    height: 0,
  });

  const getPosition = () => {
    if (elementRef.current) {
      const x = elementRef.current.offsetLeft || 0;
      const y = elementRef.current.offsetTop || 0;
      const width = elementRef.current.offsetWidth || 0;
      const height = elementRef.current.offsetHeight || 0;

      setPosition({ x, y, width, height });
    }
  };

  useEffect(() => getPosition(), [elementRef.current]);

  useEffect(() => {
    window.addEventListener('resize', getPosition, { passive: true });

    return () => window.removeEventListener('resize', getPosition);
  }, []);

  return position;
};
