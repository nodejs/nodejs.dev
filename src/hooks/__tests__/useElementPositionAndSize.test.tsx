import React, { RefObject } from 'react';
import { render } from '@testing-library/react';
import { useElementPositionAndSize } from '../useElementPositionAndSize';

type HookRendererProps<T> = {
  elementRef: RefObject<T>;
};

describe('useElementPositionAndSize', () => {
  const HookRenderer = <T extends HTMLElement>({
    elementRef,
  }: HookRendererProps<T>): JSX.Element => {
    const result = useElementPositionAndSize(elementRef);

    return <>{JSON.stringify(result)}</>;
  };

  it('should return initial value', () => {
    const elementRef = {
      current: {} as HTMLDivElement,
    } as RefObject<HTMLDivElement>;
    const { container } = render(<HookRenderer elementRef={elementRef} />);

    expect(container).toMatchSnapshot();
  });

  it('should return calculated value', () => {
    const offset = 100;
    const elementRef = {
      current: {
        offsetLeft: offset,
        offsetTop: offset,
        offsetWidth: offset,
        offsetHeight: offset,
      } as HTMLDivElement,
    } as RefObject<HTMLDivElement>;
    const { container } = render(<HookRenderer elementRef={elementRef} />);

    expect(container).toMatchSnapshot();
  });
});
