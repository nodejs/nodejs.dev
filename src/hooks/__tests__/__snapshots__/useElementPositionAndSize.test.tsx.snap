// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useElementPositionAndSize should return calculated value 1`] = `
<div>
  {"x":100,"y":100,"width":100,"height":100}
</div>
`;

exports[`useElementPositionAndSize should return initial value 1`] = `
<div>
  {"x":0,"y":0,"width":0,"height":0}
</div>
`;
